replica/database.hh

```diff
@@ -409,7 +409,7 @@ private:
     sstables::compaction_strategy _compaction_strategy;
     // TODO: Still holds a single compaction group, meaning all sstables are eligible to be compacted with one another. Soon, a table
     //  will be able to hold more than one group.
-    std::unique_ptr<compaction_group> _compaction_group;
+    lw_shared_ptr<compaction_group> _compaction_group;
```

Comment:
Side note, we need all async operations that depend on token mapping to hold on to the effective_replication_map shared instance they are based on.  Then, when there would be a concurrent operation that changes the mapping of tokens either to nodes or locally between compaction groups / tablets, I think we should use some kind of a barrier so to drain any operations that use the soon-to-be-deprecated e_r_m, and maybe consider aborting some of them that are not essential), then apply the updated e_r_m safely.
Some of the operations that were aborted could start over with the new e_r_m, e.g. cleanup compaction.

Reply:
Isn't it important to allow cleanup completion to avoid data resurrection?

Ex:

- tablet t1 owns token [A, B]
- t1 is split. t1 owns A. t2 owns B. but both owns sstable storing [A, B]
- B is deleted on t2. tombstone is soon purged with shadowed data.
- cleanup starts running on t1 and t2.
- before cleanup finishes, tablet t1 and t2 are merged again into t1.
- t1 regains ownership of B, and ressurects B.

perhaps we could have versioned replication_map, which ongoing cleanup can use to discard data that's not supposed to be there. so we can have incremental (serialized) transformations on the data as topology changes. the new received data, as part of split, won't be touched by ongoing cleanup anyway, only the old data.

Reply:
yes, similar to node ops today where decommission and remove node need to wait until cleanup that followed bootstarp, tablet merging will need to block until cleanup following splis is done.

Reply:
We already have a mechanism for this: read and write phased barriers in the table object.